### ISSUE TITLE
srt streaming: unable to stream video to nimble, makito server

### ISSUE THREAD

**Issue by boxerab** [obsproject/obs-studio#2454]
1. streaming works with VLC and wowza
1. video streaming not working with makito or nimble, audio stream is working
1. problem still occurs if sending muxed output from srt plugin 
   over udp rather than srt, so muxing is likely culprit
1. ffmpeg custom recording to url will work with nimble and makito
1. stand-alone ffmpeg works intermittently with makito - not always working

**Comment by boxerab** [comment]
cc @pkviet @JamyDev

**Comment by boxerab** [comment]
Related PR: #1748

**Comment by jamesweech** [comment]
I have provided a login to a makito decoder to @pkviet 

If anyone else needs access let me know. It supports multiple inputs so can assign one to each.

It's a Makito X Decoder on the latest firmware. (2.4.0-37)

**Comment by pkviet** [comment]
on both nimble server and makito decoder, the audio stream works; the issue is in the video stream. If I use the custom ffmpeg record output, video works on nimble server and makito.

**Comment by boxerab** [comment]
> on both nimble server and makito decoder, the audio stream works; the issue is in the video stream. If I use the custom ffmpeg record output, video works on nimble server and makito.

So, video + audio both work for nimble and makito if ffmpeg record output is used ?

**Comment by pkviet** [comment]
see https://github.com/obsproject/obs-studio/pull/2458
Makito seems to work now after these fixes.
But not nimble streamer.

**Comment by boxerab** [comment]
> see #2458
> Makito seems to work now after these fixes.
> But not nimble streamer.

Your PR changes audio encoder settings, but the problem you were seeing was with video. Interesting.

**Comment by pkviet** [comment]
> > see #2458
> > Makito seems to work now after these fixes.
> > But not nimble streamer.
> 
> Your PR changes audio encoder settings, but the problem you were seeing was with video. Interesting.

yeah it's very odd.
So on the makito decoder, we've delved a bit more and it's more complicated actually.
The makito can ingest srt and then there's a decoder which can push to a stream or to SDI/HDMI.
Before the fix, the pushed stream wouldn't be decoded for the video (only audio).
The SDI/HDMI was not working and it still does not.
So I'm betting the PES for video is malformed and lacks info. The muxing probably lacks some step.
The fact vlc and wowza decode the mpegts is probably because they are more lax to mpegts compliance to the spec.
So this does require more digging with the help of mpegts analyzers ...

**Comment by boxerab** [comment]
> The SDI/HDMI was not working and it still does not.
> So I'm betting the PES for video is malformed and lacks info. The muxing probably lacks some step.
> The fact vlc and wowza decode the mpegts is probably because they are more lax to mpegts compliance to the spec.

Interesting, as we pretty much let ffmpeg handle the packetization of the stream. So, perhaps we are missing an ffmpeg setting.

**Comment by pkviet** [comment]
Some more info, after checking with some mpegts analyzers, the PES info for the video packets seems missing. I have the impression the video headers do not contain the sps pps.

**Comment by boxerab** [comment]
> Some more info: after checking with some mpegts analyzers, the PES info for the video packets seems missing. I have the impression the video headers do not contain the sps pps.

Yes, I believe I only send the very first sps and pps, and skip the rest. Clients like ffmpeg, gstreamer and VLC are happy with this arrangement. But, apparently other servers are bit stricter.

**Comment by pkviet** [comment]
i've tried resending the video header before keyframes, but it makes no difference ...

**Comment by boxerab** [comment]
@pkviet please could you post those `tdsuck` dumps of the ts headers ? It would give some more context to the issue.

**Comment by yverbin** [comment]
I also faced with this problem. Just try to transmux mpegts received from obs srt muxer into hls using ffmpeg:
```
ffmpeg -i saved.ts -c copy -f hls -hls_time 1 -hls_list_size 0 index.m3u8
```
and you will see next errrors while opening second and subsequent segments:
```
ffprobe -i index1.ts
...
[h264 @ 0x7ff8ff028c00] decode_slice_header error
[h264 @ 0x7ff8ff028c00] no frame!
[h264 @ 0x7ff8ff028c00] non-existing PPS 0 referenced
    Last message repeated 1 times
[h264 @ 0x7ff8ff028c00] decode_slice_header error
[h264 @ 0x7ff8ff028c00] no frame!
[h264 @ 0x7ff8ff028c00] non-existing PPS 0 referenced
...
[mpegts @ 0x7ff8ff002a00] decoding for stream 0 failed
```
PAT/PMT packets looks ok.
OBS 25.0.7 MAC

**Comment by pkviet** [comment]
@yverbin if you know how to compile, you can test these fixes : 
- https://github.com/pkviet/obs-studio/tree/srtfix3
- or this PR: https://github.com/obsproject/obs-studio/pull/2665
Another thing to test is to set bframes =0 ; this is not a real fix but it might let it work.

**Comment by WizardCM** [comment]
Should this Issue be closed since #2458 and #2665 were merged, or do we want to wait for #3083 (and any future SRT fix PRs)? I would prefer to limit the scope of this Issue as best we can. @pkviet @boxerab

**Comment by pkviet** [comment]
can be closed
